torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1150,11 +1158,101 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   // threads dying due to aborted communicator and raising a SIGABRT
   std::string abortReason = c10::str("Process Group destroyed on rank ", rank_);
   abort(abortReason);
+
+  // We need to wait for abort to finish before we can safely shut down
+  // heartbeat monitoring thread.
+  terminateHeartbeatMonitorThread_.store(true);
+  monitorWakeUpCV_.notify_one();
+#ifdef ENABLE_NCCL_ERROR_CHECKING
+  if (ncclHeartbeatMonitorThread_.joinable()) {
+    ncclHeartbeatMonitorThread_.join();
+  }
+#endif
+}
+
+void dump_debugging_info() {
+  LOG(ERROR)
+      << "No PGNCCL's watchdog heartbeat detected, so we are dumping debug info.";
+  if (parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0) > 0) {
+    // TODO: Find the right and proper way to dump the debug info.
+    // We cannot print out debugging info directly.
+    LOG(ERROR) << "nccl_trace: " << dump_nccl_trace();
+  }
+}
+
+void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+  // Logging with `FATAL`, after errMsg printed, it calls `std::abort()`
+  // to terminate the program execution.
+  LOG(FATAL) << errMsg;
+}
+
+void ProcessGroupNCCL::heartbeatMonitor() {
+  uint64_t heartBeatCounter = 0ULL;
+  while (true) {
+    // This won't have any lock since this lock is only used here.
+    // Please be aware that mutex `monitorMutex_` should not be used
+    // somewhere else to avoid the deadlock.
+    std::unique_lock<std::mutex> lock(monitorMutex_);
+    if (monitorWakeUpCV_.wait_for(
+            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+              return terminateHeartbeatMonitorThread_.load();
+            })) {
+      // For the normal complete or user interception, monitorWakeUpCV_
+      // will get notified, we early return and exit heartbeatMonitor.
+      return;
+    } else {
+      auto heartbeat = heartbeat_;
+      if (heartbeat > heartBeatCounter) {
+        heartBeatCounter = heartbeat;
+      } else {
+        // No heartbeat increase detected and timeout.
+        break;
+      }
+    }
+  }
+
+  // In the timeout case and we now only dump the flight recorder
+  // to std::out. Down the road, if we have more complicated or blocking
+  // operations, we might need to use a side thread to do it.
+  dump_debugging_info();
+
+  // Create a error message reported from MonitorThread, so
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ",
+      rank_,
+      "] NCCL monitor thread timeout. Basically, this could ",
+      "be a system error and please file a bug to pytorch.");
```

Comment:
I don't think we can confidently conclude that this is a system error. 
And since this is likely not caused by a PyTorch-internal error (we've done our code right), it would be unnecessary for user to file a bug against PyTorch.

Reply:
I think a plainer explanation would suffice, something like:
"ProcessGroupNCCL's watchdog thread is stuck when cleaning or aborting collective work. This can be due to CUDA or NCCL calls being unexpectedly blocking, especially when your program enters a deadlock state. Please refer to PyTorch Distributed's document for investigating the potential deadlock."

Reply:
i think i disagree.  

we should probably write our code such that even if nccl hangs, our code does not hang.  this implies any state where our watchdog hangs "Becuase of nccl" is also our bug.

at this time, are we confident our watchdog can ONLY hang if it's a cuda/nccl bug?

also, do we want to discourage or encourage users filing an issue if there is a hang?  I prefer we know about the problem at least, so we can get it fixed

Reply:
This is not any libraries' bug, I'd call it limitation. That's why I said "CUDA or NCCL calls being unexpectedly blocking, especially when your program enters a deadlock state." The cause of deadlock state can be manyfold, such user code mistake, system error, GIL being held somewhere else, etc.

Reply:
oh, I just mention "due to CUDA or NCCL calls being unexpectedly blocking, especially when your program enters a deadlock state" and remove the system error to avoid confusion. I am open to better suggestions.